massa-models/src/block_id.rs

```diff
@@ -86,74 +136,132 @@ impl FromStr for BlockId {
                     .into_vec()
                     .map_err(|_| ModelsError::BlockIdParseError)?;
                 let u64_deserializer = U64VarIntDeserializer::new(Included(0), Included(u64::MAX));
-                let (rest, _version) = u64_deserializer
+                let (rest, version) = u64_deserializer
                     .deserialize::<DeserializeError>(&decoded_bs58_check[..])
                     .map_err(|_| ModelsError::BlockIdParseError)?;
-                Ok(BlockId(Hash::from_bytes(
-                    rest.try_into()
-                        .map_err(|_| ModelsError::BlockIdParseError)?,
-                )))
+                match version {
+                    0 => Ok(BlockIdVariant!["0"](BlockId!["0"](Hash::from_bytes(
+                        rest.try_into()
+                            .map_err(|_| ModelsError::BlockIdParseError)?,
+                    )))),
+                    _ => Err(ModelsError::BlockIdParseError),
+                }
             }
             _ => Err(ModelsError::BlockIdParseError),
         }
     }
 }
 
-impl BlockId {
-    /// block id to bytes
-    pub fn to_bytes(&self) -> &[u8; BLOCK_ID_SIZE_BYTES] {
-        self.0.to_bytes()
-    }
-
-    /// block id into bytes
-    pub fn into_bytes(self) -> [u8; BLOCK_ID_SIZE_BYTES] {
-        self.0.into_bytes()
-    }
-
-    /// block id from bytes
-    pub fn from_bytes(data: &[u8; BLOCK_ID_SIZE_BYTES]) -> BlockId {
-        BlockId(Hash::from_bytes(data))
-    }
-
-    /// first bit of the hashed block id
-    pub fn get_first_bit(&self) -> bool {
-        self.to_bytes()[0] >> 7 == 1
+#[transition::impl_version(versions("0"))]
+impl FromStr for BlockId {
+    type Err = ModelsError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut chars = s.chars();
+        match chars.next() {
+            Some(prefix) if prefix == BLOCKID_PREFIX => {
+                let data = chars.collect::<String>();
+                let decoded_bs58_check = bs58::decode(data)
+                    .with_check(None)
+                    .into_vec()
+                    .map_err(|_| ModelsError::BlockIdParseError)?;
+                let u64_deserializer = U64VarIntDeserializer::new(Included(0), Included(u64::MAX));
+                let (rest, version) = u64_deserializer
+                    .deserialize::<DeserializeError>(&decoded_bs58_check[..])
+                    .map_err(|_| ModelsError::BlockIdParseError)?;
+                match version {
+                    0 => Ok(Self(Hash::from_bytes(
+                        rest.try_into()
+                            .map_err(|_| ModelsError::BlockIdParseError)?,
+                    ))),
+                    _ => Err(ModelsError::BlockIdParseError),
+                }
+            }
+            _ => Err(ModelsError::BlockIdParseError),
+        }
     }
 }
```

Comment:
This is not the correct way to do this imo, the `FromStr` impl should only be decoding the string before sending it the object's deserializer. Here we duplicate code and increase the chances of having mismatching behaviour somewhere in the future (happened before). Also goes for the base `BlockId` implementation.